source/adapters/level_zero/command_buffer.cpp

```diff
@@ -753,17 +756,43 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
   ur_event_handle_t RetEvent{};
   // Create a command-list to signal RetEvent on completion
   ur_command_list_ptr_t SignalCommandList{};
-  if (Event) {
-    UR_CALL(Queue->Context->getAvailableCommandList(Queue, SignalCommandList,
-                                                    false, false));
+  UR_CALL(Queue->Context->getAvailableCommandList(Queue, SignalCommandList,
+                                                  false, false));
+  // Reset the wait-event for the UR command-buffer that is signalled when its
+  // submission dependencies have been satisfied.
+  ZE2UR_CALL(zeCommandListAppendEventReset,
+             (SignalCommandList->first, CommandBuffer->WaitEvent->ZeEvent));
 
+  if (Event) {
     UR_CALL(createEventAndAssociateQueue(Queue, &RetEvent,
                                          UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP,
                                          SignalCommandList, false));
 
-    ZE2UR_CALL(zeCommandListAppendBarrier,
-               (SignalCommandList->first, RetEvent->ZeEvent, 1,
-                &(CommandBuffer->SignalEvent->ZeEvent)));
+    if ((Queue->Properties & UR_QUEUE_FLAG_PROFILING_ENABLE)) {
+      // Multiple submissions of a command buffer implies that we need to save
+      // the event timestamps before resubmiting the command buffer. We
+      // therefore copy the these timestamps in a dedicated USM memory section
+      // before completing the command buffer execution, and then attach this
+      // memory to the event returned to users to allow to allow the profiling
+      // engine to recover these timestamps.
+      command_buffer_profiling_t *Profiling = new command_buffer_profiling_t();
+
+      Profiling->NumEvents = WaitEventList.size();
+      Profiling->Timestamps =
+          new ze_kernel_timestamp_result_t[Profiling->NumEvents];
+
+      ZE2UR_CALL(zeCommandListAppendQueryKernelTimestamps,
+                 (SignalCommandList->first, WaitEventList.size(),
+                  WaitEventList.data(), (void *)Profiling->Timestamps, 0,
+                  RetEvent->ZeEvent, 1,
+                  &(CommandBuffer->SignalEvent->ZeEvent)));
+
+      RetEvent->CommandData = static_cast<void *>(Profiling);
```

Comment:
so in the before we were adding always a barrier, but now only if PROFILING is enabled. why don't we need the barrier on this branch?

Reply:
The barrier was used to signal the end of this specific execution of the CommandBuffer.
The same CommandBuffer can be executed multiple times. A first barrier is added in the main command-list. https://github.com/oneapi-src/unified-runtime/pull/1035/files#diff-ab021f9451ccb6dd66e3df853d0e123b2f898233a1f609b1e698bef4f6011c71R455
This barrier waits for all commands to complete and signals its completion with `SignalEvent`.
So this barrier is the effective one. 
But `SignalEvent` is common to all executions of the CommandBuffer. 
We therefore add a second barrier in the suffix command-list that waits for `SignalEvent` and triggers a other signal, exclusive to this specific execution of the CommandBuffer. And this exclusive event is returned to the runtime.
Now, if profiling is enabled, we directly used `zeCommandListAppendQueryKernelTimestamps` to do the job since this command accepts events to wait for, and allows signaling its completion with an event.